service/storage_service.cc

```diff
@@ -471,28 +471,24 @@ void storage_service::maybe_start_sys_dist_ks() {
     _sys_dist_ks.invoke_on_all(&db::system_distributed_keyspace::start).get();
 }
 
-/* If we're not a seed node or there are seeds other than us,
- * wait until we contact any of them and learn what tokens are used by other nodes in the cluster,
- * with a timeout.
+/* Wait until we retrieve the tokens of all seeds in the cluster, with a timeout.
  *
- * Returns true if we're the only seed or we've managed to learn the other nodes' tokens before timeouting.
+ * Returns true if we've managed to learn all seeds' tokens before timeouting.
  */
 // Run in seastar::async context.
-static bool wait_for_other_tokens(const locator::token_metadata& tm,
-        const std::set<inet_address>& seeds, inet_address my_ep, abort_source& as) {
-    if (seeds.size() == 1 && seeds.count(my_ep)) {
-        return true;
-    }
-
+static bool wait_for_seed_tokens(const locator::token_metadata& tm,
+        const std::set<inet_address>& seeds, abort_source& as) {
     int retries = 30;
     while (retries--) {
-        auto& tok_to_ep = tm.get_token_to_endpoint();
-        if (std::any_of(tok_to_ep.begin(), tok_to_ep.end(),
-                [my_ep] (const std::pair<token, inet_address>& p) { return p.second != my_ep; })) {
+        if (std::all_of(seeds.begin(), seeds.end(), [&tok_to_ep = tm.get_token_to_endpoint()] (const inet_address& s) {
```

Comment:
Isn't this quadratic? I guess the number of seed nodes shouldn't be big but maybe it's better to sort both collections and do the check linearly on them.

Reply:
I assumed that `seeds` is of constant size in practice. At worst its size can be equal to the total number of nodes, and the biggest Scylla clusters out there have ~100 nodes so it won't be a problem.

Reply:
There are Cassandra deployments with 100k nodes so I don't think it's wise to assume there will always be a small number here.

Reply:
You think one would specify every node as seed?
But okay, it's not a problem, I'll make a `set` out of `seeds`.

Reply:
I don't think that but I think that if some number is beyond our control it's good to be defensive.

Reply:
No wait, this doesn't make sense. I'll have to make a `set` of `inet_address`es containing the nodes.